Fix for Issue #4136, MAGETWO-53440

vasilii-b opened the pull request:
### Description
Full description in #4136.
✔️ __Added calendar initialization for Conditional Rules when a rule is created for the 1st time__

For the condition type date the datepicker was not initialized when creating new conditions.

This PR aims to solve the issue and do not allow the user to write anything to the input.

### Fixed Issues (if relevant)
Issue described in #4136

### Manual testing scenarios
Described in #4136

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
Updated Magento_Rule rules.js file

Added calendar initializaton for Conditional Rules when a rule is created for the 1st time

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=14485) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=14485) before we can accept your contribution.<br/><hr/>**Vasilii Burlacu** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

magento-engcom-team commented:
Hi @orlangur, thank you for the review.
[ENGCOM-1205](https://jira.corp.magento.com/browse/ENGCOM-1205) has been created to process this Pull Request

magento-engcom-team commented:
Hi @orlangur, thank you for the review.
[ENGCOM-1205](https://jira.corp.magento.com/browse/ENGCOM-1205) has been created to process this Pull Request

rogyar commented:
Hi @vasilii-b. There's still a possibility to write text in the datepicker input. Check the following scenario:

![magetwo 2018-04-02](https://user-images.githubusercontent.com/2148959/43991139-7aca4b36-9d66-11e8-9d7d-7505f09c6ee5.gif)

vasilii-b commented:
Thank you, @rogyar . Will have a look on that.

Commit:
Disabled autocomplete for datepicker (calendar) input field

vasilii-b commented:
@rogyar, a fix was added for the provided scenario. Thank you!

VladimirZaets commented on lib/web/mage/calendar.js:
```diff
@@ -377,6 +377,9 @@
                 .text('') // Remove jQuery UI datepicker generated image
                 .append('<span>' + pickerButtonText + '</span>');
+
+            $(element).attr('autocomplete', 'off');
```
Hi @vasilii-b. Thanks for the collaboration. This fix removed autocomplete in all scenarios. Are you sure that is the right approach?

vasilii-b replied:
Hi @VladimirZaets,
here is a good point where we can have a long discussion, I believe.
Yes, I intended to disable the autocomplete for all scenarios and here is why.

It all started from the thing that the datepicker goal is to provide the user a possibility to select the date, not to write it by hand. The input I believe should be informative only, because the input is the date picker itself. I'd suggest making the datepicker input readonly at all.

Secondly, from the autocomplete there may come values in different format than is needed by the datepicker.
Let's say on **Website A** there is input with name `date` that has format `DD/MM/YYYY`, on **Website B** there is an input with the same name (`date`) but the format is `MM/DD/YYY`. So on **Website A** user set his desired date and when comes on **Website B** browser suggests him the previous date (from Website A) - which is wrong because of the date format.

Please let me know your thoughts on this.
Thank you!

VladimirZaets replied:
I will as product manager, because it more related to the produce manager that to developer. 

Anyway, I think we should add the configuration option to manage this situation and provide a choice to using this component in each case

vasilii-b replied:
@VladimirZaets sounds reasonable. I'll add the additional configuration.
Thank you!

sidolov commented:
Hi @vasilii-b , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

vasilii-b commented:
Hi @sidolov,
Sorry for letting this happening. I've replied to @VladimirZaets and waiting for his thoughts on that matter.
Thank you!

Commit:
magento/magento2:#4136 - Widget condition with unexpected character not preventing from saving

- Restricted user to change date input value
- Open datepicker popup once user click on the date to be changed

vasilii-b commented:
Hi @VladimirZaets,
I've found a different place, more specific for changes I've added. Also, the user interaction is changed: user can change the date only by using datepicker - input is not more available for direct editing:
![2018-11-08 18 00 28](https://user-images.githubusercontent.com/13456702/48210600-4d7a9b80-e380-11e8-945f-c47404fab2b1.gif)
Please let me know your thoughts.
Thanks!

vasilii-b commented:
@VladimirZaets, please let me know if newly changes are fine for you. Thanks!

Commit:
magento/magento2:#4136 - Widget condition with unexpected character not preventing from saving

- Added autocomplete option for calendar.js

vasilii-b commented:
Hey, what's the status of this one ? Accepted or need some more attention ? Thank you!

magento-engcom-team commented:
Hi @vasilii-b. Thank you for your contribution. 
We will aim to release these changes as part of 2.2.8. 
Please check the release notes for final confirmation.

Please, consider to port this solution to 2.3 release line.
You may use [Porting tool](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#porting-tool) to port commits automatically.

vasilii-b commented:
I'll make forward port by myself. @vishal go away
